compiler/util/exprAnalysis.cpp

```diff
@@ -0,0 +1,241 @@
+#include "baseAST.h"
+#include "stlUtil.h"
+#include "stmt.h"
+#include "exprAnalysis.h"
+
+std::map<Expr*,bool> safeExprCache;
+std::map<FnSymbol*,bool> globalManipFuncCache;
+
+/*
+ * Returns true if `e` has no side effects. Checked side effects are:
+ *  - Read/write to a global
+ *  - Is/containsessential primitive
+ *  - If it's a function call has ref arguments
+ */
+bool exprHasNoSideEffects(Expr* e) {
+  if(safeExprCache.count(e) > 0) {
+    return safeExprCache[e];
+  }
+  if(CallExpr* ce = toCallExpr(e)) {
+    if(!ce->isPrimitive()) {
+      FnSymbol* fnSym = ce->theFnSymbol();
+
+      //const bool cachedGlobalManip = globalManipFuncCache.count(fnSym);
+      const bool cachedGlobalManip = isRegisteredGlobalManip(fnSym);
+
+      if(cachedGlobalManip) {
+        if(globalManipFuncCache[fnSym]) {
+          return false;
+        }
+      }
+      for_formals(formal, fnSym) {
+        if(isReferenceType(formal->typeInfo())) {
+          safeExprCache[e] = false;
+          return false;
+        }
+      }
+
+      std::vector<BaseAST*> asts;
+      collect_asts(fnSym->body, asts);
+      if(!cachedGlobalManip){
+        for_vector(BaseAST, ast, asts) {
+          if (SymExpr* se = toSymExpr(ast)) {
+            Symbol* var = se->var;
+
+            if(!var->isImmediate() &&  isGlobal(var)){
+              safeExprCache[e] = false;
+              globalManipFuncCache[fnSym] = true;
+              return false;
+            }
+          }
+        }
+      }
+
+      //this else if lazily marks a def to be immovable if the
+      //function body has a call to another user function
+      //this can be enhanced by going recursive, but it might be a bit overkill
+      for_vector(BaseAST, ast, asts) {
+        if(CallExpr* ce = toCallExpr(ast)) {
+          if(!isNonEssentialPrimitive(ce)) {
+            safeExprCache[e] = false;
+            return false;
+          }
+        }
+      }
+      //it shouldn't be touching any globals at this point
+      globalManipFuncCache[fnSym] = false;
+    }
+    else {
+      //primitive
+      //if(ce->primitive->isEssential){
+      if(! isSafePrimitive(ce)){
+        safeExprCache[e] = false;
+        return false;
+      }
+      else {
+        //here we have a non essential primitive. But consider following issue:
+        //
+        // tmp = (atomic_read() == 5);
+        //
+        // where `==` is non-essential yet one of its children has side effects
+
+        //I had implemented the following recursion to prevent the issue , but
+        //then realized that it doesn't cause any tests to fail(and I had thought
+        //this would have been a widespread case). Maybe I was wrong?. Further
+        //this has performance impact on compilation time
+
+        /*
+        Expr* e = ce->argList.first();
+        while(e) {
+          if(!isExprSafeForReorder(e)) {//is this recursion safe?
+            return false;
+          }
+          e = e->next;
+        }
+        */
```

Comment:
I think it's reasonable to remove this comment. I don't think you need to worry about the case it's describing with my current understanding of how you're structuring denormalize. However, if you keep it this way, I think it's important for the comments in the exprAnalysis.h to indicate that this analysis does not take into account subexpressions (so PRIM_MOVE, x, call sideEffectingFn() won't even think about what sideEffectingFn does - just what the outermost expr does).